src/contexts/AzureMapVectorTileSourceProvider.tsx

```diff
@@ -0,0 +1,40 @@
+import atlas from 'azure-maps-control'
+import React, { useContext, useState } from 'react'
+import { useCheckRef } from '../hooks/useCheckRef'
+import { DataSourceType, IAzureDataSourceStatefulProviderProps, IAzureMapLayerEventType, IAzureMapsContextProps, IAzureMapSourceEventType, IAzureVectorTileSourceStatefulProviderProps, MapType } from '../types'
+import { AzureMapDataSourceRawProvider as Provider } from './AzureMapDataSourceContext'
+import { AzureMapsContext } from './AzureMapContext'
+
+/**
+ * @param id datasource identifier
+ * @param children layer providers representing datasource layers
+ * @param options vector tile datasource options: see atlas.VectorTileSourceOptions
+ * @param events a object containing sourceadded, sourceremoved event handlers
+ */
+const AzureMapVectorTileSourceStatefulProvider = ({
+  id,
+  children,
+  options,
+  events,
+}: IAzureVectorTileSourceStatefulProviderProps) => {
+  const [dataSourceRef] = useState<atlas.source.VectorTileSource>(new atlas.source.VectorTileSource(id, options))
+  const { mapRef } = useContext<IAzureMapsContextProps>(AzureMapsContext)
+  useCheckRef<MapType, DataSourceType>(mapRef, dataSourceRef, (mref, dref) => {
+    Object.entries(events || {}).filter(([_, handler]) => handler !== undefined).forEach(([eventType, handler]) => {
```

Comment:
Maybe we can a little simplify that to for example:
```
    for (const eventType in events || {}) {
      if(events[eventType]) {
              mref.events.add(eventType as IAzureMapSourceEventType, dref, events[eventType] as (e: atlas.source.Source) => void)

      }
    } 
```
For me it will look better.
What you think @ambientlight @msasinowski?

Reply:
@psrednicki: I personally don't have any preference for functional vs imperative style, should I just to your suggested approach above?

Reply:
For me, my version looks more clearly and is easier to understand.